### ISSUE TITLE
major change suggestions

### ISSUE THREAD

**Issue by SeanTolstoyevski** [samtupy/nvgt#150]
Hello Sam and all developers,

Even though this might not be the right place I want to thank you for this great project. I'm sure it will spark creativity in many people.

I have a few suggestions to make NVGT more dynamic and aligned with modern technologies. I'll explain how these suggestions simplify things, what problems they solve and what potential issues they can prevent in the future.

If my suggestions are accepted I can submit prs for these features. I can work on all of these suggestions; these are things I know how to do.

1. Using vcpkg as a Package Manager

Pros:

*   This allows NVGT to be compiled easily on many different platforms (ex: arm64, mingw etc.). CMake and vcpkg handle the whole process.
*   It makes updating packages much easier.
*   Users don't need to manually compile packages or configure anything. They can start developing quickly.
*   You gain access to a large package archive. When needed these can easily be added as dependencies to the project or replaced with other packages. Currently library files are distributed manually in the project. I don't think this is a good approach.
*   It also reduces extra files in the project. We have a lot of third-party packages and this would remove them.

Cons:

*   vcpkg compiles all the packages from scratch the first time and this can take a long time. However, in later uses, it will use the already compiled packages, so there won't be a compilation step.
*   As far as I know it's not compatible with scons. The project's build system would need to be moved to CMake or we would need to write CMake scripts in parallel. I can take care of this.
*   Some CMake knowledge is required to use vcpkg. I don't know how common this is among blind game developers but it's pretty (almost) standard in the industry.

2. Static Code Analysis

It's hard to give a detailed explanation on this topic, but here is the gist of it.
The project is writing   violation of many accepted C++ rules. This coding and writing style could cause unexpected segmentation faults or other errors later on. The simplest example of this is not using `nullptr` instead of the `NULL` constant.
These parts of the project should be analyzed and fixed with static code analysis tools.
I can't say that this is mandatory, but this coding style is definitely not healthy. If you show the repo to anyone who knows C and C++, they will tell you the same thing. I don't want to disrespect your hard work and I don't have that intention, but this coding style seems very problematic.

My suggestion is to integrate clang-tidy into the project. We can review and fix the results. If my suggestion is accepted I can take care of this as well.

This integration will greatly help us modernize the project later on. We would have a more solid codebase with many rules applied.

3. Architectural Issues (In my opinion)

This is a very subjective topic, but I don't think the architecture is healthy.
Before talking about this, I need to know if this is an actual concern or if it's just a problem for me
and everyone else is used to the project.
If we are open to discussion about this I will list the architectural issues and propose possible solutions.

Thanks again for this amazing work.

**Comment by ethindp** [comment]
Going suggestion by suggestion:

1. This probably isn't going to happen for a while. vcpkg does not integrate well with non-cmake build systems, and we use SCons for our build system. The current build system could definitely use some refinement and improvement, but vcpkg would be annoying to integrate as seemlessly as it might for cmake.
2. I have already run clang-tidy several times on the repository. Using `nullptr` instead of `NULL` has identical behavior regardless of which is used; the only significant difference is that `nullptr` results in `nullptr_t` and not an integral value (typically `0`). The repository is a mix of older and newer C++ coding practices. This is not uncommon in large C++ code bases, and isn't always necessarily a problem. Clang-tidy is already integrated into the project; you just need to run it yourself after executing the appropriate build rule to generate the compilation database. That said, I am curious which rules you would consider applying, because clang-tidy has many rules that are either not applicable or conflict with others, and being too aggressive can cause even "modern" practices to be flagged as "problems" and can therefore generate false positives as to what needs to be fixed. I have in the past executed clang-tidy on the repository and fixed actual problems, though it is far too easy to blindly apply it's advice and end up breaking things instead of fixing them, so judicious and careful application is warranted if we do want to re-run it again.
3. You have provided virtually no detail here. We are obviously open to improvements and suggestions, but without you first telling us what improvements you believe we should make architecturally I cannot provide any feedback here.

**Comment by SeanTolstoyevski** [comment]
> non-cmake build systems

can you give me more detail to me?

> Using nullptr instead of NULL has identical behavior regardless of which is used


`NULL` can be implicitly converted to integer types and pointer types which may lead to ambiguity.

`nullptr` is implicitly convertible to any pointer type but not to integer types

I don't see any sense in discussing null and nullptr specifically. I would say your arguments are not correct.

I would like to say that I partially agree with your other sentences. clang-tidy can only be a reference, we should definitely not think that this is how it should be.

> You have provided virtually no detail here

I will not write anything without working on the last item.

I will share my opinions on this issue when I examine the possible architectural problems in the repo in detail. i want to show with examples.

**Comment by ethindp** [comment]
> can you give me more detail to me?

We use SCons, not CMake. VCPKG typically only comes with build system integrations for a few build systems; SCons isn't one of them. We could probably hack it but I question the utility other than to make building on MacOS and Linux a bit easier.

> NULL can be implicitly converted to integer types and pointer types which may lead to ambiguity.
> nullptr is implicitly convertible to any pointer type but not to integer types
> I don't see any sense in discussing null and nullptr specifically. I would say your arguments are not correct.

This does not appear to be accurate. According to cppreference, `NULL` is equivalent to `nullptr`. Thus, unless the three major implementations differ from this general detail, `NULL` is identical to `nullptr`.

Regardless we are debating semantics here, because at the end of the day `NULL` and `nullptr` are identical in terms of outcome. It may be worth switching the codebase to use `nullptr`, but the codebase does include dependencies which need to remain untouched.

> I will not write anything without working on the last item.
I will share my opinions on this issue when I examine the possible architectural problems in the repo in detail. i want to show with examples.

I mean, I'm not sure what architectural issues would exist, and identifying this in a comprehensive manner would probably require a significant audit of the code. Things are quite often rewritten or changed for a variety of reasons, and technical debt here may be one of those acceptable evils.

**Comment by SeanTolstoyevski** [comment]
sir, @ethindp .

You're saying that essentially, `nullptr` and `NULL` are basically the same thing.

If that's your point, I don't think it's worth discussing further with you because it's very clear that your modern C++ knowledge is lacking or incorrect. I won't explain in detail here because I don't think this is the right place, but I'll add a few key points. I'd also say you're misinterpreting the cppref source. Please provide a link to the relevant page.
To stay on topic, I won't give another answer in the context of `nullptr` vs. `NULL`. What I wrote was just one example. There are many other problematic things like this. For example, overly complex `if-else` branches or many places that could be coded more effectively using modern C++ features.
Please consider what I've written constructively. These are all suggestions. If you think my suggestions aren't suitable for the project or don't align with the current roadmap, you can close the issue.

Not using `nullptr` has several consequences:

1. Function overload resolution errors.
2. Unintended potential implicit conversions.
3. The standard library has functions with `std::nullptr_t` parameters. Passing the `NULL` macro instead of `nullptr` to them will cause syntax errors.

If you accept CMake and vcpkg support, I can add it. Users who want to can compile with CMake and others can use SCons.
Once we create a useful CMake structure, maintaining the CMake script will be quite easy.

**Comment by ethindp** [comment]
> You're saying that essentially, `nullptr` and `NULL` are basically the same thing.

Yes, because `NULL` is an implementation-defined null-pointer value. There is nothing in the standard that says `NULL` *must* be `0`.

> If that's your point, I don't think it's worth discussing further with you because it's very clear that your modern C++ knowledge is lacking or incorrect. I won't explain in detail here because I don't think this is the right place, but I'll add a few key points. I'd also say you're misinterpreting the cppref source. Please provide a link to the relevant page.

Uh. No. Please go look at the `nullptr` cppreference page. Although it lists `NULL` as just being a macro to `nullptr` as a "possible implementation," it would not surprise me if existing implementations do this in C++11/14/17/20 mode.

> To stay on topic, I won't give another answer in the context of nullptr vs. NULL. What I wrote was just one example. There are many other problematic things like this. For example, overly complex if-else branches or many places that could be coded more effectively using modern C++ features.

Might you provide an example of this? I do know of some instances where they could be rewritten to use switch-case, as one example, but these are going to be rewritten anyway so trying to "modernize" them at this point would be, well, pointless.

I am not objecting to modernization, nor am I trying to be obstructive; however, you didn't provide examples like this in your original comment, only simple things like `NULL` being used in place of `nullptr`, hence us briefly being tied up in semantic matters. (For reference, I am very on top of C++ evolution, and use modern features such as algorithms and ranges extensively.) Modernization has some significant challenges. For one, we must be careful what we modernize and how, because we compile for five platforms (Windows, MacOS, Linux, and Android, but iOS will be added sometime in the future), and all of these targets have differing levels of C++ implementation compliance. It is far too easy to "modernize" a feature, and one of the following outcomes then occurs:

1. The resulting "modernized" version is broken or otherwise is not at feature parity with it's original code, and introduces significantly greater problems than it resolves (this happened when I attempted to modernize part of the sound mixer code in an earlier set of PRs I submitted); or
2. The code compiles on windows, and your Linux/WSL setup, but then turns out that MacOS does not implement the feature, and neither does Android (this happened with the change of some code to use `std::from_chars` for parsing floats/ints, and `std::format`, for example). We have run into this particular issue several times requiring us to revert the implementation to the prior one; as an example, this happened with my most recent PR with the Noise implementation, which is still in progress, where I needed to modify the library to not use ranges because the Android target does not support them.


In general we have found that modernization of code is not always worth it or practical. It may temporarily beautify the code, until we must revert it because a compiler toolchain setup doesn't support it, and we do not wish to introduce unnecessary dependencies to "fix" the problem.

Another issue to using vcpkg, as you asked about in your first suggestion, is that not all the dependencies we use are in vcpkg to begin with. Thus, our current build setup. It would not surprise me if this was a lot more common than even I think in the SWE industry. Currently, we use SDL3 and Steam audio, two examples that are not in vcpkg (and given the complexity of building steam audio it may never be added). Replacing those components is definitely not an option due to how tightly they are integrated into the engine.

> Please consider what I've written constructively. These are all suggestions. If you think my suggestions aren't suitable for the project or don't align with the current roadmap, you can close the issue.

Again, I am not trying to deliberately be annoying, though I am perhaps being too pedantic. I am just a contributor and don't have the power to close comments, and wouldn't want to even if I could. I apologize if anything I said came off as non-constructive.

> If you accept CMake and vcpkg support, I can add it. Users who want to can compile with CMake and others can use SCons.
> Once we create a useful CMake structure, maintaining the CMake script will be quite easy.

I don't believe this would necessarily be worthwhile, for the sole reason that SCons is so deeply integrated (and does quite a bit more than just building things, you can see an example in our top-level SConstruct file) that CMake can't do without the execution of external processes or other weird hacks. Plus, maintaining three build systems adds unnecessary complexity whenever the main build system changes, and we already need to maintain two build systems given that Android uses it's own. You are free to contribute a build system, if you want, provided you can enable equal feature parity with the original build system (or, at least, I'm pretty sure this is the general sentiment), but it will be entirely up to you to maintain and won't be an officially supported method. Adding a new build system, or switching build systems, has been discussed before, and is one of those "we might do it, maybe, someday" things that isn't something we'd consider doing in the near future right now. Before you go doing that though, I would advice you to deeply understand the current ones, because they can get quite complex in many places.

**Comment by SeanTolstoyevski** [comment]
> Yes, because NULL is an implementation-defined null-pointer value. There is nothing in the standard that says NULL must be 0.

okay.

**Comment by ethindp** [comment]
> sdl3 has been added to vcpkg.

This does not actually change anything.

> I don't know about the status of steam audio, but I don't think it's hard to compile. An overlay-port can be written and this can be much more effective than embedding everything in the project.

All I will say here is that it is very obvious you have never built steam audio. Steam audio has a unique build system that would cause an overlay port to be insufficient to build it, particularly given that some of it's dependencies are not open source. They may be optional, but this does not change the fact that using them is the only way to achieve the same performance that is achievable in the official builds.

> As the dependencies of the project increase, distributing separate archives or creating instructions for the platforms will become quite problematic, but as far as I understand, you are describing the problems you are experiencing instead of considering the suggestions.

No, I am explaining the rationale for our current build system setup and why things are done in a particular manner. The problems that I described are not just ones I have experienced but ones that have affected official releases of NVGT, and that have caused PRs to have to be reverted in the past. The build instructions for the project are not complicated or convoluted in any manner unless you are building on a non-Ubuntu system, and that can be alleviated by just updating the build_linux.sh script to take into account differing Linux distributions or including more of the binary dependencies in lindev.tar.gz. And all of this is before we consider Android.

The rest of your post consists of personal attacks and digs, and this is the second time you have engaged in such behavior. I have been perfectly constructive in this discussion, even if one part (the `NULL`/`nullptr` distinction) was something I was incorrect about. Just because you get push-back to suggestions does not indicate that my knowledge on any given subject is "lacking" or "incorrect", and my knowledge on C++ is not even relevant to this discussion anyway. As I stated, the majority of your suggestions are either not something that we can implement at this time (vcpkg), something that has already been done (clang-tidy), something that has proven to be very troublesome (modernization of the code base across five build targets with varying levels of C++ support), or something that I cannot comment on because you have provided absolutely no details and instead have hand-waved them away (architectural problems). You also only provided one example of non-modern code that was actually of any substance. I have asked you for more examples of code that isn't modern and you have provided nothing. In other words, the only actual suggestions you have provided throughout this discussion are (1) using vcpkg and (2) replacing `NULL` with `nullptr`.

If you continue engaging in personal attacks towards me, I shall have nothing more to say on this matter, and such behavior will not be tolerated here, as it is both toxic and non-constructive and will get us nowhere. Other contributors/maintainers are just as likely to reiterate the exact same issues I have already explained to you in regards to that which we cannot immediately adopt due to compiler constraints (like modernizing code).

**Comment by samtupy** [comment]
Hello,

Thanks for the complements about the project.

As to your suggestions:

1. vcpkg/cmake
To start with, I will say that what I agree with is finding a way for people to run one or 2 commands to quickly build the windev.zip package for NVGT. People should not need to do a bunch of work if they don't want to download prebuilt libraries.
What I disagree with is going all the way and switching the project's entire build system. This has a few more issues than you've probably considered, such  as:
* Most importantly everyone who wants to write an NVGT plugin then needs to learn CMake, right now they hardly need to know python. More generally how can the plugin system remain as structured as it currently is if this happens?
* Would the user folder in NVGT's repo be able to exist as it does with the SConstruct build system?
* I feel like our usage of cmake would be advanced, particularly involving the management of stubs. For example after nvgt_windows.bin gets `*compiled*,` the first 2 bytes of the PE header in the stubs are replaced with NV instead of MZ to avoid an unnecessary windows defender scan that slows down NVGT script compilation when the stub is copied. Other things I know CMake at least has some facilities for handling is the generation of our vwersion.cpp file etc, but this is still more of an advanced case and for me at least requires more learning. The biggest thing that makes me even consider CMake is it's propensity for generating files for other build systems E. someone wanted a visual studio project and cmake might be able to make one. But considering our postbuild commands issue I just described, would such a CMake project with various custom commands and actions translate cleanly to a visual studio etc project anyway?
* most generally, I don't know CMake. I plan to learn it similar to how you might have planned to take your least favorite class in school and at that time we might end up with a CMakeListts file, maybe, for those who want it, but at the moment I am uncomfortable switching to a build system I don't personally know how to use. Especially with the number of older tutorials that could cause me to learn CMake wrongly, I've not been looking forward to this. In other words, I feel like I'd be switching to CMake less because CMake is actually good, but more because everything uses it and so thus NVGT stands out if it doesn't. This does not feel like a positive reason to switch build systems for me. If someone agrees to maintain a CMake build system for NVGT though, I'd of course accept such  a thing. As it stands though, I think my fear is the loss of control and confidence in what I'm doing. Right now I know python. If I want to modify a file post-build, I create a python file object like I've been doing for years. If I want to do that with CMake, rather than actually getting to work I instead will need to learn CMake's inline code files, custom command execution or other similar mechanisms that I fear would make my development process very slow and frustrating for a while.
* regarding vcpkg directly, what if we want to use either a modified package or the development version of a package, like Angelscript's master branch? A super good example of this is SDL on Android. For accessibility reasons, I needed to modify some source code in SDL (both the c and android layer) to get everything working right now. DO I plan to share my fixes with SDL as pull requests at some point once SDL3 gets more stable? Why certainly! But at present, using the SDL3 on vcpkg might not work for reasons that are very not obvious, for example only 2 or so buttons will show in an Android message box with the official SDL where as I fixed it in our SDL by making the buttons render vertically.

What sounds appealing for me is doing something like what steam audio did. They have a script called get_dependencies.py that downloads third party libraries for that project, which then  allows it's build to succeed. In other words, I'd love it if we had a script that you can run which might harness vcpkg/conan/etc to download everything there and other repositories to fetch anything not in vcpkg to create a platdev folder which then allows the SConstruct build to succeed. It seems like this would solve the core of the problem, which is that unless you want to use the provided windev.zip file, you'd be spending hours building dependencies yourself which is quite undesirable and too nonstandard even for my tastes.

2. clang-tidy
OK, so I'm going to start by saying that the way I learned c++ personally was in a very constrained and antiquated environment with c++03 and not even so much as a printf or std::cout function. No, I do not know all of c++20 (nor did I ever claim to) and am still actively learning. If you are going to treat me like you did @ethindp because of that fact (very uncool by the way), this conversation is over and this issue will be closed. This project compiles with c++20 to give us access to the c++20 features we need, not as an attempt or declaration to create a purely c++20 conformant project.

To be very clear, I am very interested in fixing any memory leaks, seg faults, or other stylistic issues that are certainly or probably causing the code to become unstable. And though I will of course begin coding with every modern practice I learn over time, I am not interested in making the project perfectly c++20 compliant just for the sake of making it compliant.

Lets take this nullptr example you gave.

The problem is if 2 functions exist `void func(int argument)` and `void func(int* argument)`, calling `func(NULL)` would be ambiguous and not good because NULL might cast to integer 0 and call the non-pointer version of the function. I realize that. The solution is to call func(nullptr) instead.

Yes, in many instances if I bypass a NULL in the code I'll convert it to nullptr quick, have been doing that for months now. But frankly, there is no difference between the lines `void* ptr = NULL;` and `void* ptr = nullptr;` those are not ambiguous at all and will always evaluate to the same thing, the resulting ptr will be null. In such instances, I'm not interested in dropping everything I'm doing to switch instances of NULL to nullptr that have no ambiguity attached to them, and this lack of ambiguity in many instances is I believe exactly what @ethindp was talking about. By all means if I'm incorrect about this assertion, feel free to point me at a resource that explains it better than I understand.

Generally speaking, there is a lot going on with NVGT right now behind the scenes E. actively switching from bass to miniaudio, adding encryption to the core pack, finally adding tone synth based on the waveform APIs in miniaudio etc etc. It seems like it would be a huge distraction not to mention an inefficiency to focus on clang-tidy at this current project stage when new untested code will be in the pipeline days after doing so. Would it not be better to wait for heavy development to halt first then to polish all of that, as opposed to stylizing code as more is very actively being written?

3. Architectural Issues
Considering on the minor end that you've provided no detail here, but more because what we're doing seems to work for most users of NVGT, I'm going to go on a whim and say this is probably an opinion thing on your part rather than an actual problem.

Certainly, there are things we plan to do over time to make things easier for everyone, for example using github artifacts in our build workflow instead of an ftp protected by GH secrets so that forkers can build the project, nightly builds, making Android builds use more SCons and less ndk-build/gradle thus finishing the integration of the plugin system, finding a single-click method of building all the project's dependencies as described above, writing more thorough test cases for the project etc. Furthermore as this is my first large project released on Github, I would certainly expect that I'm not doing everything fully to standard.

I need to find a careful balance though between doing what I must learn is standard and doing what I already know works in an attempt to make sure that actual project development continues. One concept I keep repeating to many people who bring up such things is mental intake bandwidth. I can only learn so much at one time here and must often fall back to doing what I'm already confident in rather than learning the next newest or most standard thing right now, writing my own documentation generator for NVGT instead of finding an existing option is a great example of this. Currently all of my learning bandwidth is being exhausted by consuming the miniaudio library.

In the end, if you wish to contribute a CMake based build system I think many people would appreciate you for it so long as you are willing to maintain it. If you want to contribute some cleanup to the code it is of course very appreciated, so long as nothing breaks in the process (it's crazy how much of the c++20 standard simply isn't fully implemented on some platforms). If you know of specific bugs in the code that have a high potential of causing a crash, leak or other issue due to a malpractice, by all means please report them!

Hopefully this makes our side of things a bit more clear. I am not opposed to standardizing or modernizing things about NVGT, so long as it doesn't become an unneeded stressor, so long as it's not overdone, so long as nothing breaks in obscure ways and so long as project development doesn't need to halt for months to make room for it.

Thanks for the suggestions.

**Comment by ethindp** [comment]
As I noted earlier as well, clang-tidy is already integrated. It's not an end-to-end integration (i.e., you can't just run one scons command and it runs clang-tidy), but it is possible to run it on the entire repository already at this time, and I will occasionally run it to see what it has to say. But as @samtupy and I discovered in our most-recent run, it will also happily send us hunting ghosts and not actual problems because it is incapable of analyzing the full context of how something is invoked or used, hence the reason we must apply it judiciously and carefully (and hence why I have yet to see about us setting up a clang-tidy lint workflow or something -- the output would be insane since not everything it will lint is *our* code but code of dependencies, among other things).
With respect to the first issue, I am all for figuring out an automated build pipeline for building dependencies as well. However, incorporating it directly into the build process is just not going to happen: the explosion of build times and disk space usage is just not going to fly for many people. Hence why we use the current approach, which is, in effect, a binary cache of built dependencies, but manually generated instead of automatically done.
